astropy/io/fits/hdu/hdulist.py

```diff
@@ -194,25 +273,30 @@ def __getitem__(self, key):
             hdus = super(HDUList, self).__getitem__(key)
             return HDUList(hdus)
 
-        idx = self.index_of(key)
-        return super(HDUList, self).__getitem__(idx)
+        # Originally this used recursion, but hypothetically an HDU with
+        # a very large number of HDUs could blow the stack, so use a loop
+        # instead
+        return self._try_while_unread_hdus(super(HDUList, self).__getitem__,
```

Comment:
I think the problem might actually be in either one of `__setitem__` or whatever is happening on `enumerate(hdulist)`.  The error that it's getting upon writing the file is coming from here:

https://github.com/astropy/astropy/blob/9f003473fbbd647fcf3a3c198d9165217f828e9a/astropy/io/fits/hdu/hdulist.py#L1182

and the error in the log says

```
E               VerifyError: 
E               Verification reported errors:
E               HDUList's element 2 is not an extension HDU.
E               Note: astropy.io.fits uses zero-based indexing
```

Note *zero-based indexing*.  At no point in the test is a third HDU (element 2) added to the HDUList, yet enumerate is finding a third element that, whatever the heck it is, is not an instance of `ExtensionHDU`.